From 7aa663846d3d6d02ce876563f04b589d4eb31d30 Mon Sep 17 00:00:00 2001
From: Han Wang <416810799@qq.com>
Date: Sun, 23 Dec 2018 18:18:45 +0100
Subject: [PATCH 03/11] lineage-sdk: Handle database downgrading

 * This is ususally messed up when doing bringing ups. Handle this
   by re-creating databases instead of crash.

Change-Id: Ic98c14e08af3fc7dd65e3f241b25f4f3c7835348
---
 .../lineagesettings/LineageDatabaseHelper.java    | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/packages/LineageSettingsProvider/src/org/lineageos/lineagesettings/LineageDatabaseHelper.java b/packages/LineageSettingsProvider/src/org/lineageos/lineagesettings/LineageDatabaseHelper.java
index 9589b4d..9a849d0 100644
--- a/packages/LineageSettingsProvider/src/org/lineageos/lineagesettings/LineageDatabaseHelper.java
+++ b/packages/LineageSettingsProvider/src/org/lineageos/lineagesettings/LineageDatabaseHelper.java
@@ -382,6 +382,21 @@ public class LineageDatabaseHelper extends SQLiteOpenHelper{
         }
     }
 
+    @Override
+    public void onDowngrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+        Log.w(TAG, "Cannot downgrade db. Old version: " + oldVersion + ", new version: "
+                + newVersion + ". Must wipe the lineage settings provider.");
+
+        dropDbTable(db, LineageTableNames.TABLE_SYSTEM);
+        dropDbTable(db, LineageTableNames.TABLE_SECURE);
+
+        if (mUserHandle == UserHandle.USER_OWNER) {
+            dropDbTable(db, LineageTableNames.TABLE_GLOBAL);
+        }
+
+        onCreate(db);
+    }
+
     private void moveSettingsToNewTable(SQLiteDatabase db,
                                         String sourceTable, String destTable,
                                         String[] settingsToMove, boolean doIgnore) {
-- 
2.17.1

